File: 1201-1300/1233-Remove Sub-Folders from the Filesystem/1233-Remove Sub-Folders from the Filesystem.cpp
class TrieNode {
public:
    TrieNode() {
        memset(next, NULL, sizeof(next));
        isEnd = false;
    };

    ~TrieNode() {
        for (TrieNode* p : next) {
            if (p) delete p;
        }
    }

    TrieNode* next[27];
    bool isEnd;
};

class Solution {
public:
    vector<string> removeSubfolders(vector<string>& folder) {
        sort(folder.begin(), folder.end(), [](string &s1, string &s2){return s1.size() < s2.size();});
        TrieNode* root = new TrieNode();
        vector<string> result;
        for (string f : folder) {
            TrieNode* curr = root;
            bool subfolder = false;
            for (int i = 0; i < f.size(); ++i) {
                int index = (f[i] == '/') ? 26 : f[i] - 'a';
                if (curr->next[index] == nullptr) {
                    curr->next[index] = new TrieNode();
                }
                curr = curr->next[index];
                if (curr->isEnd && f[i + 1] == '/') {
                    subfolder = true;
                    break;
                }
            }

            if (!subfolder) {
                result.push_back(f);
                curr->isEnd = true;
            }
        }

        return result;
    }
};

File: 1201-1300/1233-Remove Sub-Folders from the Filesystem/1233-Remove Sub-Folders from the Filesystem.py
import collections
class TrieNode:
    def __init__(self):
        self.next = collections.defaultdict(TrieNode)
        self.end = False

class Solution:
    def removeSubfolders(self, folder: List[str]) -> List[str]:
        folder.sort(key = lambda x : len(x))
        result = []
        root = TrieNode()
        for f in folder:
            curr = root
            subfolder = False
            for i, c in enumerate(f):
                curr = curr.next[c]
                if curr.end and f[i + 1] == '/':
                    subfolder = True
                    break
            
            if not subfolder:
                result.append(f)
                curr.end = True

        return result
